Implement generic Table for basic CRUD operations & concurrent safe dequeuing

VojtechVitek opened the pull request:
## Concurrent dequeue/update pattern
- Using a shared lock using PostgreSQL's `FOR UPDATE SKIP LOCKED` pattern
- Concurrent dequeue/process example: https://github.com/goware/pgkit/blob/d081629296b0c4afd49887d47d9fb45795329cf6/tests/table_test.go#L193-L200
- Dequeue code: https://github.com/goware/pgkit/blob/d081629296b0c4afd49887d47d9fb45795329cf6/tests/tables_test.go#L24-L47
- Worker processing code: https://github.com/goware/pgkit/blob/d081629296b0c4afd49887d47d9fb45795329cf6/tests/worker_test.go#L24-L64

## Data models with strongly typed `*data.Record` type (using generics):
- CRUD usage:  https://github.com/goware/pgkit/blob/d081629296b0c4afd49887d47d9fb45795329cf6/tests/table_test.go#L24-L45
- DB transaction across multiple tables: https://github.com/goware/pgkit/blob/d081629296b0c4afd49887d47d9fb45795329cf6/tests/table_test.go#L63-L83

VojtechVitek committed:
Implement generic Table for basic CRUD operations

Co-authored-by: David Sedlacek <david.sedlacek@golang.cz>

VojtechVitek committed:
Remove automatic 'deleted_at NULL' condition

VojtechVitek committed:
Implement .LockForUpdate() using PostgreSQL's FOR UPDATE SKIP LOCKED pattern

VojtechVitek committed:
Fix generic ID

VojtechVitek committed:
Fix naming of generic types

VojtechVitek committed:
Add tests for simple CRUD, complex transactions and LockForUpdate()

VojtechVitek committed:
Fix TestRecordsWithJSONStruct test, since schema changed

VojtechVitek committed:
Fix LockForUpdate test

VojtechVitek committed:
Don't rely on wg.Go(), a feature from Go 1.25

VojtechVitek committed:
LockForUpdate that can pass transaction to update fn

VojtechVitek committed:
Refactor LockForUpdate() to reuse tx if possible

VojtechVitek committed:
Simplify data models further

.Save() - variadic arg instead of .Save() and .SaveAll()
.List() - renamed from .GetAll()
.Get()  - renamed from .GetOne()

.LockForUpdates() - renamed from .LockForUpdate()
.LockForUpdate()  - renamed from .LockOneForUpdate()

VojtechVitek committed:
Improve tests for async processing

VojtechVitek committed:
Tests: Implement in-memory worker pattern via simple WaitGroup

VojtechVitek committed:
A better "dequeue" abstraction defined on reviews table

VojtechVitek committed:
Rename GetByIDs() to ListByIDs()

david-littlefarmer commented on table.go:
```diff
@@ -0,0 +1,261 @@
+	}
+
+	return nil
+}
```
I would flip logic around, Do the iteration in `Save()` and put the save query to `save()`.
Currently if i save more records it's calling `Save()` -> `saveAll()`iterate and call -> `Save()`

VojtechVitek replied:
I think we can optimize `saveAll()` with batch inserts eventually -- so that's why it's flipped it like this.

Otherwise, I'd agree 👍

david-littlefarmer reviewed (approved):
One comment, looks good, it would be nice to also have paginated methods.

VojtechVitek commented:
Exactly, great point about the pagination 👍 I'd love to see both LIMIT/OFFSET and cursor-based pagination defined as a base data model methods. This needs some more thoughts, so I skipped it for now.

shunkakinoki commented on tests/schema_test.go:
```diff
@@ -11,19 +12,85 @@ type Account struct {
 	Name      string    `db:"name"`
 	Disabled  bool      `db:"disabled"`
 	CreatedAt time.Time `db:"created_at,omitempty"` // ,omitempty will rely on postgres DEFAULT
+	UpdatedAt time.Time `db:"created_at,omitempty"` // ,omitempty will rely on postgres DEFAULT
```
```suggestion
	UpdatedAt time.Time `db:"updated_at,omitempty"` // ,omitempty will rely on postgres DEFAULT
```

shunkakinoki commented on table.go:
```diff
@@ -0,0 +1,261 @@
+		Limit(1).
+		OrderBy(orderBy...)
+
+	if err := t.Query.GetOne(ctx, q, record); err != nil {
```
`fmt.Errorf` wrap?

VojtechVitek committed:
Fix updated_at field, thanks @shunkakinoki

shunkakinoki commented on table.go:
```diff
@@ -0,0 +1,261 @@
+}
+
+// Count returns the number of matching records.
+func (t *Table[T, PT, IDT]) Count(ctx context.Context, where sq.Sqlizer) (uint64, error) {
```
Make `where sq.Sqlizer` a non-nil  clause so that it doesn't check for all conditions?

VojtechVitek replied:
Thank you. I think it's a totally valid point to worry about performance of the `nil` case, which effectively queries all table data:

```sql
SELECT COUNT(1) FROM table
```

Imho, I think we should keep it as is & provide a warning about this in the comment.

And also, I think it'd be great to provide a faster alternative that returns "estimated count" via a query such as
```
-- Planner’s estimate of live tuples in the table. Updated by VACUUM, ANALYZE, certain DDL.  
SELECT reltuples::BIGINT AS estimated_count 
FROM pg_class 
WHERE relname = 'table_name';
```

or

```
-- Estimated live rows tracked by the statistics collector (monitoring view) for each table.  
SELECT n_live_tup
  FROM pg_stat_user_tables
 WHERE relname = 'table_name';
```

^^ need to test these first

pkieltyka replied:
what is the need for us to have a Count() method on the table..? I am nervous for offering this feature to developers, because it means they'll use it, and counting a table is one of the slowest things you can do.. can we just remove this method entirely..?

VojtechVitek replied:
Imho there's nothing wrong with `SELECT COUNT(1)` as long as we hit the index properly, e.g.:

```
SELECT COUNT(1) FROM jobs WHERE status = 2;
```

But I hear you, this should never get "abused" for wrong purposes, such as counting number of total pages for LIMIT OFFSET pagination.

VojtechVitek replied:
@klaidliadon  @david-littlefarmer what do you use `COUNT(1)` or `COUNT(*)` for in builder and marketplace-api? I see several use of `COUNT` in these codebases. Are there any legit use-cases?

david-littlefarmer replied:
Yes, we are counting the orders (listings/offers) so we can show the numbers on top of the page on FE. How many listings are there as total count and as filtered count.

klaidliadon replied:
In the Builder we use count using foreign keys and/or indexed values.
There are a few exceptions for tables that have a really low row count.

VojtechVitek committed:
PR feedback: Improve error annotations

VojtechVitek committed:
Fix tests

pkieltyka commented on table.go:
```diff
@@ -0,0 +1,271 @@
+}
+
+// saveAll saves multiple records.
+// TODO: This function can be likely optimized to use a batch insert query.
```
yes, please make this a batch insert or this is going to be misused and get us into performance troubles

VojtechVitek replied:
@david-littlefarmer will work on this 👍

c2h5oh commented on table.go:
```diff
@@ -0,0 +1,271 @@
+		return fmt.Errorf("select for update skip locked: %w", err)
+	}
+
+	updateFn(records)
```
update function should return error

c2h5oh replied:
update function should also have access to pgTx - in case it has to do do other related db changes

VojtechVitek replied:
Good comments, thank you!

### 1. updateFn() should return error

I used to have `updateFn() error` signature in the past but it didn't prove necessary. The semantics of `LOCK FOR UPDATE` is that we always want to finish the update and commit back to the database. For example, if we encounter an error during dequeuing or processing we can set `status=queued` or `status=failed`. But from the transaction perspective, nothing went wrong and we save the result back to the DB anyway.

At least that's the majority of use-cases I can think of.

### 2. pgTx transaction

However, what you're describing is also possible -- but you'd need to wrap the whole `LockForUpdate()` method into your own transaction manually:

```go
err := db.BeginTx(ctx, func(tx *Database) error { 
	err := tx.Reviews.LockForUpdates(ctx, where, orderBy, limit, func(reviews []*Review) { 
 		now := time.Now().UTC() 
 		for _, review := range reviews { 
 			review.Status = ReviewStatusProcessing 
 			review.ProcessedAt = &now 
 		} 
 		dequeued = reviews 
 	})
	if err != nil { 
 		return nil, fmt.Errorf("lock for updates: %w", err) 
 	} 
 	
 	// now you can update other tables too within the same transaction
	return tx.AuditLogs.Save(&Log{})
 }
```

david-littlefarmer committed:
Save multiple (#33)

* save multiple

* pr comments

* return error instead of nil on nil record

klaidliadon committed:
Add iterator method for accounts and update tests

klaidliadon committed:
Rename types for clarity

klaidliadon committed:
add back truncateAllTables

klaidliadon committed:
Add IDColumn to table context and enhance WithTx tests

klaidliadon committed:
Add pagination support to Table with ListPaged and WithPaginator methods

klaidliadon committed:
feat: add Insert and Update methods in Table

klaidliadon committed:
feat: add RestoreByID and export timestamp interfaces

- Add RestoreByID method that clears DeletedAt by passing zero time
- Export HasSetCreatedAt, HasSetUpdatedAt, HasSetDeletedAt interfaces
  with godoc explaining the contract for each lifecycle hook
- Update SetDeletedAt implementations to treat zero time as restore (nil)

klaidliadon committed:
fix: LockForUpdates missing return, saveAll hardcoded ID column, Iter missing rows.Err, saveOne missing SetCreatedAt

- LockForUpdates: add missing return after existing-tx branch to prevent
  double execution of updateFn via a second transaction
- saveAll: use t.IDColumn instead of hardcoded "id" for update WHERE clause
- Iter: check rows.Err() after iteration loop to surface driver errors
- saveOne: call SetCreatedAt on insert path (consistent with saveAll)
- Use min() builtin for chunk bounds

klaidliadon committed:
fix(tests): race conditions in TestLockForUpdates and wrong ListByIDs slice

- Move wg.Add(1) before goroutine dispatch in ProcessReview to prevent
  worker.Wait() returning early before all goroutines register
- Replace shared ids[][]uint64 slice (data race across goroutines) with
  per-worker local slices merged under mutex
- Fix articleIDs slice: make([]uint64, len) + append produced leading zeros,
  changed to make([]uint64, 0, len)

klaidliadon committed:
fix: ListPaged nil page panic

Normalize nil page before passing to PrepareQuery/PrepareResult.
PrepareQuery creates a local &Page{} but the caller's pointer stays nil,
causing PrepareResult to panic on page.More assignment.

klaidliadon committed:
fix: LockForUpdates validate/timestamp, zero-value Paginator, ListPaged ordering

- lockForUpdatesWithTx: call Validate() and SetUpdatedAt() on records
  after updateFn, matching Insert/Update/Save behavior
- Page.SetDefaults: fall back to DefaultPageSize/MaxPageSize when
  PaginatorSettings has zero values (fixes zero-value Paginator capping
  page size to 0)
- ListPaged: add IDColumn fallback ordering when no sort is configured,
  ensuring deterministic pagination
- TestLockForUpdates: use assert instead of require in goroutine
  (require.FailNow is unsafe off the test goroutine)

klaidliadon committed:
fix: PrepareRaw empty ORDER BY and unbound limit/offset params

- Skip ORDER BY clause when no sort columns are configured
- Always inject pgx.NamedArgs for limit/offset even when args is empty

klaidliadon committed:
feat: Update, DeleteByID, HardDeleteByID return (bool, error)

Return true when at least one row was affected, false when zero rows
matched. Callers can now distinguish "success" from "not found" without
reimplementing methods with RowsAffected checks.

Breaking change: signatures go from error to (bool, error).

VojtechVitek commented:
Just adding this here for future reference:

I think we should document clearly that `.Count()` can have a real performance impact, as `COUNT(1)` is not necessarily a cheap SQL operation.

We could add a new `.CountEstimate()` method, which returns the estimated number of all rows in a table, which is the fast alternative to `.Count()` -- however it would not support WHERE clause -- which is a big downside.

```
-- Planner’s estimate of live tuples in the table. Updated by VACUUM, ANALYZE, certain DDL.  
SELECT reltuples::BIGINT AS estimated_count 
FROM pg_class 
WHERE relname = 'table_name';
```

or

```
-- Estimated live rows tracked by the statistics collector (monitoring view) for each table.  
SELECT n_live_tup
  FROM pg_stat_user_tables
 WHERE relname = 'table_name';
 ```

klaidliadon commented:
On top of what @VojtechVitek covered in the PR:

### `Update`, `DeleteByID`, `HardDeleteByID` now return `(bool, error)`
- `true` = at least one row affected, `false` = no match — no more reimplementing methods just to check `RowsAffected`
- **Breaking change** from the previous `error`-only signature

### New methods
- `Insert` (single + bulk with chunking)
- `RestoreByID` for un-soft-deleting
- `ListPaged` with `Paginator` support, `WithPaginator` for custom page size/max
- `WithTx` preserves paginator settings
- `Iter` for streaming records

### Fixes
- `saveAll` was hardcoding the ID column
- `saveOne` wasn't calling `SetCreatedAt`
- Race conditions in `TestLockForUpdates`

VojtechVitek commented on table.go:
```diff
@@ -0,0 +1,631 @@
+	Validate() error
+}
+
+// Table provides basic CRUD operations for database records.
```
```suggestion
// Table provides basic CRUD operations for database records.
// NOTICE: Experimental. Table and its methods are subject to change.
```

VojtechVitek committed:
NOTICE: Experimental. Table and its methods are subject to change.
